readme-parser/parse.ts

```diff
@@ -0,0 +1,318 @@
+
+const getTextFromNode = (
+	node: RootContent | PhrasingContent | undefined,
+): string | null => {
+	if (!node) {
+		return null;
+	}
+
+	if ('value' in node) {
+		return node.value;
+	}
+
+	if ('children' in node) {
+		return getTextFromNode(node.children[0]);
+	}
+
+	return null;
+};
+
+const getUrlFromNode = (
+	node: RootContent | PhrasingContent | null,
+): string | null => {
+	if (!node) {
+		return null;
+	}
+
+	if ('url' in node) {
+		return node.url;
+	}
+
+	if ('children' in node) {
+		return getUrlFromNode(node.children[0] ?? null);
+	}
+
+	return null;
+};
+
+const getHeading = (
+	rootContents: ReadonlyArray<RootContent>,
+	depth: 1 | 2 | 3,
+	name: string | null,
+): Heading | null => {
+	for (const rootContent of rootContents) {
+		if (rootContent.type !== 'heading') {
+			continue;
+		}
+
+		if (name !== null) {
+			const headerTitle = getTextFromNode(rootContent);
+
+			if (
+				!headerTitle ||
+				!noFirstLetterLowerCase(headerTitle).startsWith(
+					noFirstLetterLowerCase(name),
+				)
+			) {
+				continue;
+			}
+		}
+
+		if (rootContent.depth === depth) {
+			return rootContent;
+		}
+	}
+
+	return null;
+};
+
+const getTextByHeader = (
+	rootContents: ReadonlyArray<RootContent>,
+	heading: Heading,
+	delimiter: string,
+) => {
+	const headerIndex = rootContents.findIndex(
+		(rc) => rc.position?.start.line === heading.position?.start.line,
+	);
+	const nextHeaderIndex = rootContents.findIndex(
+		(rc) =>
+			rc.type === 'heading' &&
+			rc.position?.start.line &&
+			heading.position?.start.line &&
+			rc.position?.start.line > heading.position?.start.line &&
+			rc.depth === heading.depth,
+	);
+
+	const contentParts = rootContents.slice(
+		headerIndex + 1,
+		nextHeaderIndex > -1 ? nextHeaderIndex : undefined,
+	);
+
+	const textParts: string[] = [];
+
+	for (const rc of contentParts) {
+		if ('children' in rc) {
+			rc.children
+				.map((child) => {
+					if (child.type === 'text') {
+						return `${child.value}${delimiter}`;
+					}
+
+					if (child.type === 'listItem') {
+						return `${getUrlFromNode(
+							child.children[0] ?? null,
+						)}${delimiter}`;
+					}
+
+					if (child.type === 'link' || child.type === 'strong') {
+						return getTextFromNode(child.children[0]);
+					}
+
+					// Do not add new line after certain blocks (treated as separate AST nodes)
+					if (child.type === 'inlineCode') {
+						return child.value;
+					}
+
+					return null;
+				})
+				.forEach((child) => {
+					if (child !== null) {
+						textParts.push(child);
+					}
+				});
+		}
+
+		if ('value' in rc) {
+			if (rc.type === 'code') {
+				textParts.push(
+					`\`\`\`${rc.lang}\n${rc.value}\n\`\`\`${delimiter}`,
+				);
+			} else {
+				textParts.push(`${rc.value}${delimiter}`);
+			}
+		}
+	}
+
+	// Trim last el to remove delimiter
+	const lastElement = textParts[textParts.length - 1];
+	if (lastElement) {
+		const lastSymbol = lastElement[lastElement.length - 1];
```

Comment:
can it be regexed?

Reply:
i would argue that this is way easier to read than what i've replaced it with, but it works for me